From fc8ab5b8017c95d536a653e6f7bef38025066231 Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Mon, 8 Sep 2014 14:32:06 -0400
Subject: [scsi] hpsa: fix handling of hpsa_volume_offline return value

Message-id: <1410186734-19986-76-git-send-email-thenzl@redhat.com>
Patchwork-id: 92094
O-Subject: [RHEL6.6 e-stor PATCH 75/83] hpsa: fix handling of hpsa_volume_offline return value
Bugzilla: 1069185
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Ewan Milne <emilne@redhat.com>

Make return value an int instead of an unsigned char so that
we do not lose negative error return values.

Signed-off-by: Stephen M. Cameron <scameron@beardog.cce.hp.com>
Reviewed-by: Webb Scales <webb.scales@hp.com>
Reported-by: Dan Carpenter <dan.carpenter@oracle.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit 67955ba36eccb5733fd289839a5a5d1dfc33ccf7)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/hpsa.c b/drivers/scsi/hpsa.c
index bbd571b..202ec83 100644
--- a/drivers/scsi/hpsa.c
+++ b/drivers/scsi/hpsa.c
@@ -2549,11 +2549,11 @@ exit_failed:
 /* Determine offline status of a volume.
  * Return either:
  *  0 (not offline)
- * -1 (offline for unknown reasons)
+ *  0xff (offline for unknown reasons)
  *  # (integer code indicating one of several NOT READY states
  *     describing why a volume is to be kept offline)
  */
-static unsigned char hpsa_volume_offline(struct ctlr_info *h,
+static int hpsa_volume_offline(struct ctlr_info *h,
      unsigned char scsi3addr[])
 {
  struct CommandList *c;
@@ -2652,11 +2652,15 @@ static int hpsa_update_device_info(struct ctlr_info *h,
 
  if (this_device->devtype == TYPE_DISK &&
   is_logical_dev_addr_mode(scsi3addr)) {
+  int volume_offline;
+
   hpsa_get_raid_level(h, scsi3addr, &this_device->raid_level);
   if (h->fw_support & MISC_FW_RAID_OFFLOAD_BASIC)
    hpsa_get_ioaccel_status(h, scsi3addr, this_device);
-  this_device->volume_offline =
-   hpsa_volume_offline(h, scsi3addr);
+  volume_offline = hpsa_volume_offline(h, scsi3addr);
+  if (volume_offline < 0 || volume_offline > 0xff)
+   volume_offline = HPSA_VPD_LV_STATUS_UNSUPPORTED;
+  this_device->volume_offline = volume_offline & 0xff;
  } else {
   this_device->raid_level = RAID_UNKNOWN;
   this_device->offload_config = 0;
diff --git a/drivers/scsi/hpsa_cmd.h b/drivers/scsi/hpsa_cmd.h
index 649b463..b5125dc 100644
--- a/drivers/scsi/hpsa_cmd.h
+++ b/drivers/scsi/hpsa_cmd.h
@@ -151,7 +151,7 @@
 #define HPSA_VPD_HEADER_SZ              4
 
 /* Logical volume states */
-#define HPSA_VPD_LV_STATUS_UNSUPPORTED   -1
+#define HPSA_VPD_LV_STATUS_UNSUPPORTED   0xff
 #define HPSA_LV_OK                                      0x0
 #define HPSA_LV_UNDERGOING_ERASE   0x0F
 #define HPSA_LV_UNDERGOING_RPI    0x12
-- 
1.7.1